lib/kubernetes-deploy/ejson_secret_provisioner.rb

```diff
@@ -0,0 +1,153 @@
+# frozen_string_literal: true
+require 'json'
+require 'base64'
+require 'open3'
+require 'kubernetes-deploy/kubeclient_builder'
+require 'kubernetes-deploy/logger'
+
+module KubernetesDeploy
+  class EjsonSecretError < FatalDeploymentError
+    def initialize(msg)
+      super("Creation of Kubernetes secrets from ejson failed: #{msg}")
+    end
+  end
+
+  class EjsonSecretProvisioner
+    include KubernetesDeploy::KubeclientBuilder
+
+    MANAGEMENT_ANNOTATION = "kubernetes-deploy.shopify.io/ejson-secret".freeze
+    MANAGED_SECRET_EJSON_KEY = "kubernetes_secrets".freeze
+    EJSON_SECRETS_FILE = "secrets.kubernetes-deploy.ejson".freeze
+    EJSON_KEYS_SECRET = "ejson-keys".freeze
+
+    def initialize(namespace:, context:, template_dir:, client: nil)
+      @namespace = namespace
+      @context = context
+      @ejson_file = "#{template_dir}/#{EJSON_SECRETS_FILE}"
+      @kubeclient = client || build_v1_kubeclient(@context)
+    end
+
+    def secrets_requested?
+      File.exist?(@ejson_file)
+    end
+
+    def create_secrets
+      raise EjsonSecretError, "#{EJSON_SECRETS_FILE} not found" unless secrets_requested?
+
+      with_decrypted_ejson do |decrypted|
+        secrets = decrypted[MANAGED_SECRET_EJSON_KEY]
+        unless secrets.present?
+          KubernetesDeploy.logger.warn("#{EJSON_SECRETS_FILE} does not have key #{MANAGED_SECRET_EJSON_KEY}. No secrets will be created.")
+          return
+        end
+
+        secrets.each do |secret_name, secret_spec|
+          validate_secret_spec(secret_name, secret_spec)
+          create_or_update_secret(secret_name, secret_spec["_type"], secret_spec["data"])
+        end
+      end
+    end
+
+    def prune_managed_secrets
+      ejson_secret_names = encrypted_ejson.fetch(MANAGED_SECRET_EJSON_KEY, {}).keys
+      live_secrets = @kubeclient.get_secrets(namespace: @namespace)
+
+      live_secrets.each do |secret|
+        secret_name = secret.metadata.name
+        next unless secret.metadata.annotations.to_h.stringify_keys.key?(MANAGEMENT_ANNOTATION)
+        next if ejson_secret_names.include?(secret_name)
+
+        KubernetesDeploy.logger.info("Pruning secret #{secret_name}")
```

Comment:
I'd kinda rather that warnings stay things they should worry about when they see. If warnings are expected, I'd expect people will start ignoring them. Like they do on core deploys. Worth noting that we do not log warnings when pruning other resources (we don't even say what we pruned, though that we definitely should do).